File: backend/indicators.py
def rsi(closes: list[float], period: int = 14) -> float:
    if len(closes) < period + 1:
        return 50.0
    deltas = [closes[i] - closes[i - 1] for i in range(1, len(closes))]
    gains = [max(d, 0) for d in deltas[-period:]]
    losses = [abs(min(d, 0)) for d in deltas[-period:]]
    avg_gain = sum(gains) / period
    avg_loss = sum(losses) / period
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return round(100 - (100 / (1 + rs)), 2)


def sparkline(closes: list[float]) -> list[float]:
    return [round(c, 2) for c in closes[-24:]]


def _ema_series(values: list[float], period: int) -> list[float]:
    if not values:
        return []
    if len(values) < period:
        return [values[-1]] * len(values)
    multiplier = 2 / (period + 1)
    series: list[float] = []
    seed = sum(values[:period]) / period
    series.extend([seed] * period)
    ema_val = seed
    for price in values[period:]:
        ema_val = (price - ema_val) * multiplier + ema_val
        series.append(ema_val)
    return series


def ema(closes: list[float], period: int) -> float:
    series = _ema_series(closes, period)
    return round(series[-1], 2) if series else 0.0


def macd(closes: list[float], fast: int = 12, slow: int = 26, signal: int = 9) -> dict:
    if len(closes) < slow + signal:
        return {"macd": 0.0, "signal": 0.0, "histogram": 0.0}
    fast_series = _ema_series(closes, fast)
    slow_series = _ema_series(closes, slow)
    macd_line = [f - s for f, s in zip(fast_series, slow_series)]
    signal_series = _ema_series(macd_line, signal)
    macd_val = macd_line[-1]
    signal_val = signal_series[-1]
    return {
        "macd": round(macd_val, 4),
        "signal": round(signal_val, 4),
        "histogram": round(macd_val - signal_val, 4),
    }


def bollinger_bands(closes: list[float], period: int = 20, std_dev: float = 2.0) -> dict:
    if len(closes) < period:
        mid = closes[-1] if closes else 0.0
        return {"upper": round(mid, 2), "middle": round(mid, 2), "lower": round(mid, 2)}
    window = closes[-period:]
    middle = sum(window) / period
    variance = sum((x - middle) ** 2 for x in window) / period
    std = variance ** 0.5
    return {
        "upper": round(middle + std_dev * std, 2),
        "middle": round(middle, 2),
        "lower": round(middle - std_dev * std, 2),
    }


def compute_all(closes: list[float]) -> dict:
    macd_vals = macd(closes)
    bb = bollinger_bands(closes)
    return {
        "rsi": rsi(closes),
        "ema_9": ema(closes, 9),
        "ema_21": ema(closes, 21),
        "macd": macd_vals["macd"],
        "macd_signal": macd_vals["signal"],
        "macd_histogram": macd_vals["histogram"],
        "bb_upper": bb["upper"],
        "bb_middle": bb["middle"],
        "bb_lower": bb["lower"],
    }

File: frontend/lib/indicators.ts
import type { Indicators } from "./api";

export type CandleLike = { close: number };

/** Latest indicator values from candle closes — matches chart series exactly. */
export function snapshotFromCloses(closes: number[]): Indicators | null {
  if (closes.length < 2) return null;
  return computeAll(closes);
}

export function snapshotFromCandles(candles: CandleLike[]): Indicators | null {
  return snapshotFromCloses(candles.map((c) => c.close));
}

export function rsi(closes: number[], period = 14): number {
  if (closes.length < period + 1) return 50;
  const deltas = closes.slice(1).map((c, i) => c - closes[i]);
  const gains = deltas.slice(-period).map((d) => Math.max(d, 0));
  const losses = deltas.slice(-period).map((d) => Math.abs(Math.min(d, 0)));
  const avgGain = gains.reduce((a, b) => a + b, 0) / period;
  const avgLoss = losses.reduce((a, b) => a + b, 0) / period;
  if (avgLoss === 0) return 100;
  return Math.round((100 - 100 / (1 + avgGain / avgLoss)) * 100) / 100;
}

/** RSI value at each bar (for chart series). */
export function rsiSeries(closes: number[], period = 14): number[] {
  return closes.map((_, i) => rsi(closes.slice(0, i + 1), period));
}

export function sparkline(closes: number[]): number[] {
  return closes.slice(-24).map((c) => Math.round(c * 100) / 100);
}

export function emaSeries(values: number[], period: number): number[] {
  if (!values.length) return [];
  if (values.length < period) return values.map(() => values[values.length - 1]);
  const k = 2 / (period + 1);
  const out: number[] = [];
  let ema = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
  out.push(...Array(period).fill(ema));
  for (let i = period; i < values.length; i++) {
    ema = values[i] * k + ema * (1 - k);
    out.push(ema);
  }
  return out;
}

function ema(closes: number[], period: number): number {
  const s = emaSeries(closes, period);
  return s.length ? Math.round(s[s.length - 1] * 100) / 100 : 0;
}

function macd(closes: number[]) {
  if (closes.length < 35) return { macd: 0, signal: 0, histogram: 0 };
  const fast = emaSeries(closes, 12);
  const slow = emaSeries(closes, 26);
  const line = fast.map((f, i) => f - slow[i]);
  const signal = emaSeries(line, 9);
  const macdVal = line[line.length - 1];
  const signalVal = signal[signal.length - 1];
  return {
    macd: Math.round(macdVal * 10000) / 10000,
    signal: Math.round(signalVal * 10000) / 10000,
    histogram: Math.round((macdVal - signalVal) * 10000) / 10000,
  };
}

function bollinger(closes: number[], period = 20, stdDev = 2) {
  if (closes.length < period) {
    const mid = closes[closes.length - 1] ?? 0;
    return { upper: mid, middle: mid, lower: mid };
  }
  const window = closes.slice(-period);
  const middle = window.reduce((a, b) => a + b, 0) / period;
  const variance = window.reduce((a, x) => a + (x - middle) ** 2, 0) / period;
  const std = Math.sqrt(variance);
  return {
    upper: Math.round((middle + stdDev * std) * 100) / 100,
    middle: Math.round(middle * 100) / 100,
    lower: Math.round((middle - stdDev * std) * 100) / 100,
  };
}

export function bollingerSeries(closes: number[], period = 20, stdDev = 2) {
  const upper: number[] = [];
  const middle: number[] = [];
  const lower: number[] = [];
  for (let i = 0; i < closes.length; i++) {
    const bb = bollinger(closes.slice(0, i + 1), period, stdDev);
    upper.push(bb.upper);
    middle.push(bb.middle);
    lower.push(bb.lower);
  }
  return { upper, middle, lower };
}

export function macdSeries(closes: number[]) {
  const ema12 = emaSeries(closes, 12);
  const ema26 = emaSeries(closes, 26);
  const line = ema12.map((f, i) => f - ema26[i]);
  const signal = emaSeries(line, 9);
  const histogram = line.map((v, i) => v - signal[i]);
  return { line, signal, histogram };
}

export function computeAll(closes: number[]): Indicators {
  const m = macd(closes);
  const bb = bollinger(closes);
  return {
    rsi: rsi(closes),
    ema_9: ema(closes, 9),
    ema_21: ema(closes, 21),
    macd: m.macd,
    macd_signal: m.signal,
    macd_histogram: m.histogram,
    bb_upper: bb.upper,
    bb_middle: bb.middle,
    bb_lower: bb.lower,
  };
}
